### DISCUSSION TITLE
How to set dynamic prefix to S3PickleIOManager ?

### DISCUSSION THREAD

**Discussion by LPauzies** [dagster-io/dagster#19825] [Q&A]
Hello,

I'm currently defining jobs in dagster, that are collections of assets, each of my assets possess an intermediate result that I save using S3PickleIOManager, writing into a MinIO. However, when launching the job multiple times with different configurations, my artifacts logged by the io-manager as pickle file are stored in the same place.

For example : `dagster/myjob/myasset`.

But what I would want is the following : `dagster/myjob/runid/myasset`.

Doing this, i'm sure to retrieve an intermediate state from one of the job I have launched previously, being sure it has not been overwritten.

Some code : 

```Python
### IO Managers
def get_s3_io_manager() -> S3PickleIOManager:
    return S3PickleIOManager(
        s3_resource=get_minio_s3_resource(),
        s3_bucket="myjob",
        s3_prefix="dagster",
    )
    
resources = {
    "s3_io_manager": get_s3_io_manager(),
}
```

And some code skeleton of an asset : 

```Python
@asset(
    key_prefix=["myjob"],
    name="validate_configuration",
    description="Validate configuration for myjob",
    compute_kind="python",
    group_name="myjob",
    io_manager_key="s3_io_manager",
)
def validate_configuration(
    context: AssetExecutionContext, config: InputConfiguration
) -> ValidatedInputConfiguration:
    pass
```

As you can see, I cannot pass dynamically the run_id into the s3_prefix or path of the io_manager and the S3PickleIOManager seems to not been designed for this. If you have any idea, feel free to propose !

Thanks !

**Comment by LPauzies** [comment]
Up ?

**Comment by ion-elgreco** [comment]
@LPauzies you could create a custom IO manager that accesses the `definition_metadata` to update the write path. In your asset metadata you then add {"prefix":"experiment"}

Also the OutputContext has run_id property so you could also make use of that to prefix it
